Doc/library/typing.rst

```diff
@@ -1409,49 +1459,64 @@ These can be used as types in annotations. They all support subscription using
+   3. If the return value is ``False``, the type of its argument
+      is narrowed to exclude the type inside ``TypeIs``.
 
    For example::
 
-         def is_str_list(val: list[object]) -> TypeGuard[list[str]]:
-             '''Determines whether all objects in the list are strings'''
-             return all(isinstance(x, str) for x in val)
-
-         def func1(val: list[object]):
-             if is_str_list(val):
-                 # Type of ``val`` is narrowed to ``list[str]``.
-                 print(" ".join(val))
-             else:
-                 # Type of ``val`` remains as ``list[object]``.
-                 print("Not a list of strings!")
-
-   If ``is_str_list`` is a class or instance method, then the type in
+        from typing import assert_type, final, TypeIs
+
+        class Parent: pass
+        class Child(Parent): pass
+        @final
+        class Unrelated: pass
+
+        def is_parent(val: object) -> TypeIs[Parent]:
+            return isinstance(val, Parent)
+
+        def run(arg: Child | Unrelated):
+            if is_parent(arg):
+                # Type of ``arg`` is narrowed to the intersection
+                # of ``Parent`` and ``Child``, which is equivalent to
+                # ``Child``.
+                assert_type(arg, Child)
+            else:
+                # Type of ``arg`` is narrowed to exclude ``Parent``,
+                # so only ``Unrelated`` is left.
+                assert_type(arg, Unrelated)
+
+   The type inside ``TypeIs`` must be consistent with the type of the
+   function's argument; if it is not, static type checkers will raise
+   an error.  An incorrectly written ``TypeIs`` function can lead to
+   unsound behavior in the type system; it is the user's responsibility
+   to write such functions in a type-safe manner.
+
+   If a ``TypeIs`` function is a class or instance method, then the type in
    ``TypeGuard`` maps to the type of the second parameter after ``cls`` or
    ``self``.
 
-   In short, the form ``def foo(arg: TypeA) -> TypeGuard[TypeB]: ...``,
-   means that if ``foo(arg)`` returns ``True``, then ``arg`` narrows from
-   ``TypeA`` to ``TypeB``.
+   In short, the form ``def foo(arg: TypeA) -> TypeIs[TypeB]: ...``,
+   means that if ``foo(arg)`` returns ``True``, then ``arg`` is an instance
+   of ``TypeB``, and if it returns ``False``, it is not an instance of ``TypeB``.
 
-   .. note::
-
-      ``TypeB`` need not be a narrower form of ``TypeA`` -- it can even be a
-      wider form. The main reason is to allow for things like
-      narrowing ``list[object]`` to ``list[str]`` even though the latter
-      is not a subtype of the former, since ``list`` is invariant.
-      The responsibility of writing type-safe type guards is left to the user.
-
-   ``TypeGuard`` also works with type variables.  See :pep:`647` for more details.
+   ``TypeIs`` also works with type variables.  For more information, see
+   PEP 742 (Narrowing types with ``TypeIs``).
```

Comment:
Same question here as below -- should we link to typing spec instead of PEP?

If we do link to PEP, should probably make it a real link using `:pep:` role